onnx/defs/training/defs.cc

```diff
@@ -0,0 +1,777 @@
+            "concatenated list of attribute \"xs\" and attribute \"zs\". For "
+            "example, if xs=[\"A\", \"B\"] and zs=[\"C\"], the first input is "
+            "used as the value of symbol \"A\" and the 3rd input is "
+            "substituted for all the occurrences of \"C\".",
+            "T2",
+            OpSchema::Variadic,
+            false)
+        .Output(
+            0,
+            "Output",
+            "The gradient of the tensors specified in attribute \"y\" "
+            "with respect to \"xs\". The i-th output is the gradient "
+            "of \"y\" with respect to the i-th tensor specified in the "
+            "attribute \"xs\".",
+            "T1",
+            OpSchema::Variadic,
+            false)
+        .Attr(
+            "xs",
+            "Input tensor names of the differentiated sub-graph. It "
+            "contains and only contains the necessary differentiated "
+            "inputs of a (sub-)graph. Variables (usually called "
+            "intermediate variables) which can be generated by inputs "
+            "cannot be included in this attribute.",
+            AttributeProto::STRINGS)
+        .Attr(
+            "zs",
+            "Input tensor names of the differentiated sub-graph. It "
+            "contains and only contains the necessary non-differentiated "
+            " inputs of a (sub-)graph. Variables (usually called "
+            "intermediate variables) which can be generated by inputs "
+            "cannot be included in this attribute.",
+            AttributeProto::STRINGS,
+            OPTIONAL)
+        .Attr(
+            "y",
+            "The targeted tensor. It can be viewed as the output of the "
+            "differentiated function. The attribute \"xs\" and attribute "
+            "\"zs\" are the minimal variable set that determines the value "
+            " of \"y\".",
+            AttributeProto::STRING)
+        .TypeConstraint(
+            "T1",
+            {"tensor(float16)",
+             "tensor(float)",
+             "tensor(double)"},
+            "Constrain associated inputs and outputs to floating-point tensors.")
+        .TypeConstraint(
+            "T2",
+            {"tensor(float16)",
+             "tensor(float)",
+             "tensor(double)",
+             "tensor(int8)",
+             "tensor(int16)",
+             "tensor(int32)",
+             "tensor(int64)",
+             "tensor(uint8)",
+             "tensor(uint16)",
+             "tensor(uint32)",
+             "tensor(uint64)",
+             "tensor(bool)",
+             "tensor(string)"},
+            "Allow associated inputs and outputs to be any kinds of tensors."));
+
+const char* reduction_doc =
+    "Type of reduction to apply to loss: none, sum, mean(default). "
+    "'none': the output is the loss for each sample in the batch."
+    "'sum': the output will be summed. "
+    "'mean': the sum of the output will be divided by the batch_size.";
```

Comment:
Makes sense. Since most losses are composable, we need to propose them one-by-one like what we did in the beginning. I will continue the optimizer and loss work in difference places. This PR will only contain spec-wise changes.